src/neo-vm/Types/Map.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Neo.VM.Types
+{
+    public class Map : StackItem, ICollection, IDictionary<StackItem, StackItem>
+    {
+        private readonly Dictionary<StackItem, StackItem> dictionary;
+
+        public StackItem this[StackItem key]
+        {
+            get => this.dictionary[key];
+            set => this.dictionary[key] = value;
+        }
+
+        public ICollection<StackItem> Keys => dictionary.Keys;
+        public ICollection<StackItem> Values => dictionary.Values;
+        public int Count => dictionary.Count;
+        public bool IsReadOnly => false;
+
+        bool ICollection.IsSynchronized => false;
+        object ICollection.SyncRoot => dictionary;
+
+        public Map() : this(new Dictionary<StackItem, StackItem>()) { }
+
+        public Map(Dictionary<StackItem, StackItem> value)
+        {
+            this.dictionary = value;
+        }
+
+        public void Add(StackItem key, StackItem value)
+        {
+            dictionary.Add(key, value);
+        }
+
+        void ICollection<KeyValuePair<StackItem, StackItem>>.Add(KeyValuePair<StackItem, StackItem> item)
+        {
+            dictionary.Add(item.Key, item.Value);
+        }
+
+        public void Clear()
+        {
+            dictionary.Clear();
+        }
+
+        bool ICollection<KeyValuePair<StackItem, StackItem>>.Contains(KeyValuePair<StackItem, StackItem> item)
+        {
+            return dictionary.ContainsKey(item.Key);
+        }
+
+        public bool ContainsKey(StackItem key)
+        {
+            return dictionary.ContainsKey(key);
+        }
+
+        void ICollection<KeyValuePair<StackItem, StackItem>>.CopyTo(KeyValuePair<StackItem, StackItem>[] array, int arrayIndex)
+        {
+            foreach (KeyValuePair<StackItem, StackItem> item in dictionary)
+                array[arrayIndex++] = item;
+        }
+
+        void ICollection.CopyTo(System.Array array, int index)
+        {
+            foreach (KeyValuePair<StackItem, StackItem> item in dictionary)
+                array.SetValue(item, index++);
+        }
+
+        public override bool Equals(StackItem other)
+        {
+            return false;
```

Comment:
Should this work similarly to `VMArray` `equals`? I.e. if the maps have the same key and values, they are equal?

Reply:
I think we should change `Array.Equals` to return `false` to support TypeScript/JavaScript:

```js
var a = [1, 2, 3];
var b = [1, 2, 3];
var c = a === b; //Should be false
```

And we should override `Struct.Equals` to support the `ValueType` in C#.

Reply:
Sounds good.